### PR TITLE
Swift services HA and upgrade

### PR DESCRIPTION
**Pull request by kaweue**
Added:
- swift proxy service
- implemented controller logic of upgrade and scale
- e2e tests for upgrade and scaling

### CODE REVIEW

File: pkg/apis/contrail/v1alpha1/swift_types.go [hunk 1]
```diff
@@ -7,7 +7,8 @@ import (
 // SwiftSpec defines the desired state of Swift
 // +k8s:openapi-gen=true
 type SwiftSpec struct {
-	ServiceConfiguration SwiftConfiguration `json:"serviceConfiguration"`
+	CommonConfiguration  CommonConfiguration `json:"commonConfiguration"`
+	ServiceConfiguration SwiftConfiguration  `json:"serviceConfiguration"`
 }
 
 // SwiftConfiguration is the Spec for the keystone API.
```
Comment (shifali19) [inline pkg/apis/contrail/v1alpha1/swift_types.go:14]: Swift API

File: pkg/controller/swiftproxy/swiftproxy_controller.go [hunk 2]
```diff
@@ -24,6 +24,7 @@ import (
 
 	contrail "github.com/Juniper/contrail-operator/pkg/apis/contrail/v1alpha1"
 	"github.com/Juniper/contrail-operator/pkg/certificates"
+	"github.com/Juniper/contrail-operator/pkg/label"
```
Comment (shifali19) [inline pkg/controller/swiftproxy/swiftproxy_controller.go:27]: organize imports

File: pkg/controller/swiftproxy/swiftproxy_controller.go [hunk 3]
```diff
@@ -221,7 +258,41 @@ func (r *ReconcileSwiftProxy) Reconcile(request reconcile.Request) (reconcile.Re
...
+		for i, p := range svc.Spec.Ports {
+			if p.Port == listenPort {
+				nodePort = svc.Spec.Ports[i].NodePort
+			}
+		}
+		svc.Spec.Ports = []core.ServicePort{
+			{Port: listenPort, Protocol: "TCP", NodePort: nodePort},
+		}
+		svc.Spec.Selector = swiftProxy.Labels
+		svc.Spec.Type = core.ServiceTypeLoadBalancer
```
Comment (shifali19) [inline pkg/controller/swiftproxy/swiftproxy_controller.go:292]: Why are we using LoadBalancer service type and then using ClusterIP? If it is exposed as LoadBalancer type, shouldn't it be LoadBalancerIP?

File: pkg/controller/swiftproxy/swiftproxy_controller_test.go [hunk 4]
```diff
@@ -242,15 +292,23 @@ func newExpectedDeployment(status apps.DeploymentStatus) *apps.Deployment {
 			OwnerReferences: []meta.OwnerReference{
 				{"contrail.juniper.net/v1alpha1", "SwiftProxy", "swiftproxy", "", &trueVal, &trueVal},
 			},
-			Labels: map[string]string{"SwiftProxy": "swiftproxy"},
+			Labels: map[string]string{"SwiftProxy": "swiftproxy", "contrail_manager": "SwiftProxy"},
```
Comment (shifali19) [inline pkg/controller/swiftproxy/swiftproxy_controller_test.go:295]: Do these labels persist?I think they are overwritten in controller.

File: pkg/apis/contrail/v1alpha1/swift_types.go (hunk 1 above)
Comment (shifali19) [inline pkg/apis/contrail/v1alpha1/swift_types.go:14]: Swift API.

File: pkg/controller/swiftproxy/swiftproxy_controller.go (hunk 2 above)
Comment (shifali19) [inline pkg/controller/swiftproxy/swiftproxy_controller.go:27]: Organize imports.

File: pkg/controller/swiftproxy/swiftproxy_controller.go (hunk 3 above)
Comment (shifali19) [inline pkg/controller/swiftproxy/swiftproxy_controller.go:292]: If we are exposing the service as "LoadBalancer" type, why are we using ClusterIP? Why not LoadBalancerIP?

File: pkg/controller/swiftproxy/swiftproxy_controller_test.go (hunk 4 above)
Comment (kaweue) [reply pkg/controller/swiftproxy/swiftproxy_controller_test.go:295]: This expected deployment so basically this is what is expected to be generated from swiftproxy controller

Comment (kaweue) [reply pkg/controller/swiftproxy/swiftproxy_controller_test.go:295]: This expected deployment so basically this is what is expected to be generated from swiftproxy controller

File: pkg/controller/swiftproxy/swiftproxy_controller.go (hunk 3 above)
Comment (kaweue) [reply pkg/controller/swiftproxy/swiftproxy_controller.go:292]: LoadBalancer service type in k8s is actually 3 in 1. It is ClusterIP, NodePort and LoadBalancer.
For internal communication I think it is better to use ClusterIP.
LoadBalancer IP should be used to provision "Public" endpoint in Keystone Service registry. I left a comment
```	// TODO Use public IP provided by LoadBalancer service
	// External IP might be provisioned later, so it has to be handled asynchronously```

I will address that in next PR

File: pkg/apis/contrail/v1alpha1/swift_types.go (hunk 1 above)
Comment (kaweue) [reply pkg/apis/contrail/v1alpha1/swift_types.go]: changed

File: pkg/controller/swiftproxy/swiftproxy_controller.go (hunk 2 above)
Comment (kaweue) [reply pkg/controller/swiftproxy/swiftproxy_controller.go]: organized

**Comment by kaweue** [comment]
recheck

**Comment by kaweue** [comment]
recheck

**Comment by kaweue** [comment]
recheck

**Comment by kaweue** [comment]
recheck ha

**Pull request merged**